gnovm/pkg/gnolang/store.go

```diff
@@ -73,53 +73,276 @@ type Store interface {
+}
+
+func (b txLogMap[K, V]) Get(k K) (V, bool) {
+	if bufValue, ok := b.dirty[k]; ok {
+		if bufValue.deleted {
+			var zeroV V
+			return zeroV, false
+		}
+		return bufValue.v, true
+	}
+
+	return b.source.Get(k)
+}
+
+func (b txLogMap[K, V]) Set(k K, v V) {
+	b.dirty[k] = deletable[V]{v: v}
+}
+
+func (b txLogMap[K, V]) Delete(k K) {
+	b.dirty[k] = deletable[V]{deleted: true}
+}
+
+func (b txLogMap[K, V]) Iterate() func(yield func(K, V) bool) {
+	return func(yield func(K, V) bool) {
+		b.source.Iterate()(func(k K, v V) bool {
+			if dirty, ok := b.dirty[k]; ok {
+				if dirty.deleted {
+					return true
+				}
+				return yield(k, dirty.v)
+			}
+
+			// not in dirty
+			return yield(k, v)
+		})
+		// yield for new values
+		for k, v := range b.dirty {
+			if v.deleted {
+				continue
+			}
+			_, ok := b.source.Get(k)
+			if ok {
+				continue
+			}
+			if !yield(k, v.v) {
+				break
+			}
+		}
+	}
+}
+
+type mapWrapper[K comparable, V any] map[K]V
+
+func (m mapWrapper[K, V]) Get(k K) (V, bool) {
+	v, ok := m[k]
+	return v, ok
+}
+
+func (m mapWrapper[K, V]) Set(k K, v V) {
+	m[k] = v
+}
+
+func (m mapWrapper[K, V]) Delete(k K) {
+	delete(m, k)
+}
+
+func (m mapWrapper[K, V]) Iterate() func(yield func(K, V) bool) {
+	return func(yield func(K, V) bool) {
+		for k, v := range m {
+			if !yield(k, v) {
+				return
+			}
+		}
+	}
+}
+
+type deletable[V any] struct {
+	v       V
+	deleted bool
 }
 
 func NewStore(alloc *Allocator, baseStore, iavlStore store.Store) *defaultStore {
 	ds := &defaultStore{
-		alloc:            alloc,
+		baseStore: baseStore,
+		iavlStore: iavlStore,
+		alloc:     alloc,
+
+		// cacheObjects is set; objects in the store will be copied over for any transaction.
+		cacheObjects: make(map[ObjectID]Object),
+		cacheTypes:   mapWrapper[TypeID, Type](map[TypeID]Type{}),
+		cacheNodes:   mapWrapper[Location, BlockNode](map[Location]BlockNode{}),
+
+		// store configuration
 		pkgGetter:        nil,
-		cacheObjects:     make(map[ObjectID]Object),
-		cacheTypes:       make(map[TypeID]Type),
-		cacheNodes:       make(map[Location]BlockNode),
 		cacheNativeTypes: make(map[reflect.Type]Type),
-		baseStore:        baseStore,
-		iavlStore:        iavlStore,
+		pkgInjector:      nil,
+		nativeStore:      nil,
 		go2gnoStrict:     true,
 	}
 	InitStoreCaches(ds)
 	return ds
 }
 
+// Unstable.
+// This function is used to clear the object cache every transaction.
+// It also sets a new allocator.
+func (ds *defaultStore) ClearObjectCache() {
+	ds.alloc.Reset()
+	ds.cacheObjects = make(map[ObjectID]Object) // new cache.
+	ds.opslog = nil                             // new ops log.
+	ds.SetCachePackage(Uverse())
+}
+
+// If nil baseStore and iavlStore, the baseStores are re-used.
+func (ds *defaultStore) BeginTransaction(baseStore, iavlStore store.Store) TransactionStore {
+	if baseStore == nil {
+		baseStore = ds.baseStore
+	}
+	if iavlStore == nil {
+		iavlStore = ds.iavlStore
+	}
+	ds2 := &defaultStore{
```

Comment:
Nitpick: I think we can name this better

Reply:
You mean instead of `ds2`?